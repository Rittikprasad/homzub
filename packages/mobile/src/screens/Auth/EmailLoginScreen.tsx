import React from 'react';
import { StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { IRedirectionDetails } from '@homzhub/mobile/src/services/LinkingService';
import { NavigationService } from '@homzhub/mobile/src/services/NavigationService';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { CommonSelectors } from '@homzhub/common/src/modules/common/selectors';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { LoginForm } from '@homzhub/common/src/components/organisms/LoginForm';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import {
  IEmailLoginPayload,
  ILoginFormData,
  ILoginPayload,
  LoginTypes,
} from '@homzhub/common/src/domain/repositories/interfaces';

interface IStateProps {
  isLoading: boolean;
  redirectionDetails: IRedirectionDetails;
}

interface IDispatchProps {
  login: (payload: ILoginPayload) => void;
}

type libraryProps = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.EmailLogin>;
type Props = IDispatchProps & IStateProps & libraryProps;

export class EmailLoginScreen extends React.PureComponent<Props> {
  public render(): React.ReactNode {
    const { t, isLoading } = this.props;
    return (
      <Screen
        containerStyle={styles.container}
        headerProps={{
          type: 'secondary',
          icon: icons.leftArrow,
          onIconPress: this.handleIconPress,
        }}
        pageHeaderProps={{
          contentTitle: t('auth:logInWithEmail'),
        }}
        backgroundColor={theme.colors.white}
        keyboardShouldPersistTaps
        isLoading={isLoading}
      >
        <LoginForm
          isEmailLogin
          handleForgotPassword={this.handleForgotPassword}
          onLoginSuccess={this.handleLoginSuccess}
          testID="loginForm"
        />
      </Screen>
    );
  }

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private handleForgotPassword = (): void => {
    const {
      navigation,
      route: { params },
    } = this.props;
    const onCallback = params && params.onCallback ? { onCallback: params.onCallback } : {};
    navigation.navigate(ScreensKeys.ForgotPassword, onCallback);
  };

  private handleLoginSuccess = (values: ILoginFormData): void => {
    const {
      login,
      route: { params },
    } = this.props;

    const emailLoginData: IEmailLoginPayload = {
      action: LoginTypes.EMAIL,
      payload: {
        email: values.email,
        password: values.password,
      },
    };

    const loginPayload: ILoginPayload = {
      data: emailLoginData,
      ...(params && params.onCallback && { callback: params.onCallback }),
      handleDynamicLink: this.handleDynamicLink,
    };

    login(loginPayload);
  };

  private handleDynamicLink = (): void => {
    const { redirectionDetails } = this.props;
    if (redirectionDetails.shouldRedirect && redirectionDetails.redirectionLink) {
      NavigationService.handleDynamicLinkNavigation(redirectionDetails).then();
    }
  };
}

export const mapStateToProps = (state: IState): IStateProps => {
  return {
    isLoading: UserSelector.getLoadingState(state),
    redirectionDetails: CommonSelectors.getRedirectionDetails(state) as IRedirectionDetails,
  };
};

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { login } = UserActions;
  return bindActionCreators(
    {
      login,
    },
    dispatch
  );
};

// @ts-ignore
export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation()(EmailLoginScreen));

// @ts-ignore
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
});
