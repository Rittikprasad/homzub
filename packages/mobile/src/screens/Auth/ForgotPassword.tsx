import React, { Component } from 'react';
import { StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { UserRepository } from '@homzhub/common/src/domain/repositories/UserRepository';
import { IForgotPasswordPayload } from '@homzhub/common/src/domain/repositories/interfaces';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { Text } from '@homzhub/common/src/components/atoms/Text';
import { FormButton } from '@homzhub/common/src/components/molecules/FormButton';
import { FormTextInput } from '@homzhub/common/src/components/molecules/FormTextInput';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';

interface IForgotPasswordState {
  email: string;
  isRequestDisable: boolean;
}

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.ForgotPassword>;

export class ForgotPassword extends Component<Props, IForgotPasswordState> {
  public state = {
    email: '',
    isRequestDisable: false,
  };

  public render(): React.ReactNode {
    const {
      t,
      route: { params },
    } = this.props;
    const { isRequestDisable } = this.state;
    const formData = { ...this.state };
    return (
      <Screen
        headerProps={{
          type: 'secondary',
          icon: icons.close,
          onIconPress: this.handleIconPress,
        }}
        pageHeaderProps={{
          contentTitle: t('auth:forgotPassword'),
          contentSubTitle: t('auth:forgotPasswordDescription'),
          disableDivider: true,
        }}
        backgroundColor={theme.colors.white}
      >
        <Formik onSubmit={this.onSubmit} validate={FormUtils.validate(this.formSchema)} initialValues={formData}>
          {(formProps: FormikProps<FormikValues>): React.ReactElement => (
            <>
              <FormTextInput
                formProps={formProps}
                inputType="email"
                name="email"
                label="Email"
                placeholder={t('auth:enterEmail')}
                isMandatory
                onValueChange={this.onValueChange}
              />
              <FormButton
                formProps={formProps}
                type="primary"
                title={t('auth:forgotPasswordRequestLink')}
                // @ts-ignore
                onPress={formProps.handleSubmit}
                containerStyle={styles.formButtonStyle}
                disabled={!formProps.values.email || (formProps.values.email && isRequestDisable)}
              />
            </>
          )}
        </Formik>
        {params && !params.isFromMore && (
          <Text
            type="small"
            textType="semiBold"
            style={styles.backToLoginLink}
            onPress={this.navigateToLogin}
            testID="txtLogin"
          >
            {t('auth:backToLogin')}
          </Text>
        )}
      </Screen>
    );
  }

  private onValueChange = (): void => {
    const { isRequestDisable } = this.state;
    if (isRequestDisable) {
      this.setState({ isRequestDisable: false });
    }
  };

  private onSubmit = async (formProps: IForgotPasswordState): Promise<void> => {
    const { t } = this.props;
    const { email } = formProps;
    const payload: IForgotPasswordPayload = {
      action: 'SEND_EMAIL',
      payload: {
        email,
      },
    };

    this.setState({ isRequestDisable: true });

    try {
      const emailData = await UserRepository.emailExists(email);
      if (emailData.is_exists) {
        UserRepository.resetPassword(payload)
          .then(() => {
            AlertHelper.success({ message: t('auth:resetLink') });
          })
          .catch((err) => {
            AlertHelper.error({ message: err });
          });
      } else {
        AlertHelper.error({ message: t('auth:emailNotExists') });
      }
    } catch (e) {
      AlertHelper.error({ message: e.message });
    }
  };

  public navigateToLogin = (): void => {
    const {
      navigation,
      route: { params },
    } = this.props;
    const onCallback = params && params.onCallback ? { onCallback: params.onCallback } : {};
    navigation.navigate(ScreensKeys.EmailLogin, onCallback);
  };

  public handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private formSchema = (): yup.ObjectSchema<{ email: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      email: yup.string().email(t('auth:emailValidation')).required(t('auth:emailRequired')),
    });
  };
}

export default withTranslation()(ForgotPassword);

const styles = StyleSheet.create({
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
  backToLoginLink: {
    color: theme.colors.blue,
    alignSelf: 'center',
  },
});
