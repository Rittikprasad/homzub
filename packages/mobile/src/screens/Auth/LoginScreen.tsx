import React, { Component } from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { AuthenticationGateways } from '@homzhub/mobile/src/components';
import { LoginForm } from '@homzhub/common/src/components/organisms/LoginForm';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import { NavigationScreenProps, OtpNavTypes, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { ILoginFormData } from '@homzhub/common/src/domain/repositories/interfaces';

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.Login>;

interface IOwnState {
  isLoading: boolean;
}

export class LoginScreen extends Component<Props, IOwnState> {
  public state = {
    isLoading: false,
  };

  public render(): React.ReactNode {
    const {
      t,
      navigation,
      route: { params },
    } = this.props;
    const { isLoading } = this.state;

    return (
      <Screen
        headerProps={{
          type: 'secondary',
          icon: icons.close,
          onIconPress: this.onClosePress,
        }}
        pageHeaderProps={{
          contentTitle: t('login'),
          contentSubTitle: t('auth:newAroundHere'),
          contentLink: t('auth:signup'),
          onLinkPress: this.onSignUpClicked,
        }}
        backgroundColor={theme.colors.white}
        isLoading={isLoading}
      >
        <>
          <LoginForm onLoginSuccess={this.onOtpLoginPress} testID="loginForm" />
          <AuthenticationGateways
            isFromLogin
            onEmailLogin={this.onEmailLoginPress}
            onSuccessCallback={params.onCallback}
            navigation={navigation}
            testID="socialEmailLogin"
            toggleLoading={this.toggleLoading}
          />
        </>
      </Screen>
    );
  }

  private onClosePress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private onEmailLoginPress = (): void => {
    const {
      navigation,
      route: { params },
    } = this.props;
    const onCallback = params && params.onCallback ? { onCallback: params.onCallback } : {};
    navigation.navigate(ScreensKeys.EmailLogin, onCallback);
  };

  private onSignUpClicked = (): void => {
    const {
      navigation,
      route: { params },
    } = this.props;
    const onCallback = params && params.onCallback ? { onCallback: params.onCallback } : {};
    navigation.navigate(ScreensKeys.SignUp, onCallback);
  };

  private onOtpLoginPress = (values: ILoginFormData): void => {
    const {
      navigation,
      t,
      route: { params },
    } = this.props;
    navigation.navigate(ScreensKeys.OTP, {
      type: OtpNavTypes.Login,
      title: t('auth:loginOtp') ?? '',
      countryCode: values.phone_code,
      otpSentTo: values.phone_number,
      ...(params && params.onCallback && { onCallback: params.onCallback }),
    });
  };

  private toggleLoading = (isLoading: boolean): void => {
    this.setState({ isLoading });
  };
}

export default withTranslation()(LoginScreen);
