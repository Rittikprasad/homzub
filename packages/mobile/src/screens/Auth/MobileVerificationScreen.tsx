import React, { Component } from 'react';
import { StyleSheet } from 'react-native';
import { Formik, FormikHelpers, FormikProps } from 'formik';
import * as yup from 'yup';
import { withTranslation, WithTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { UserRepository } from '@homzhub/common/src/domain/repositories/UserRepository';
import { theme } from '@homzhub/common/src/styles/theme';
import { Text } from '@homzhub/common/src/components/atoms/Text';
import { FormButton } from '@homzhub/common/src/components/molecules/FormButton';
import { FormTextInput } from '@homzhub/common/src/components/molecules/FormTextInput';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { NavigationScreenProps, OtpNavTypes, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { SocialAuthKeys } from '@homzhub/common/src/constants/SocialAuthProviders';

interface IVerificationState {
  phone: string;
  phoneCode: string;
}

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.MobileVerification>;

export class MobileVerificationScreen extends Component<Props, IVerificationState> {
  public state = {
    phone: '',
    phoneCode: '',
  };

  public render(): React.ReactNode {
    const { t } = this.props;
    const { title, message, subTitle, buttonTitle } = this.getDisplayStrings();

    return (
      <Screen
        headerProps={{
          type: 'secondary',
          onIconPress: this.handleIconPress,
        }}
        pageHeaderProps={{
          contentTitle: title,
          contentSubTitle: subTitle,
        }}
        backgroundColor={theme.colors.white}
        contentContainerStyle={styles.container}
      >
        <Text type="small" style={styles.message}>
          {message}
        </Text>
        <Formik
          initialValues={{ ...this.state }}
          onSubmit={this.onSubmit}
          validate={FormUtils.validate(this.formSchema)}
        >
          {(formProps: FormikProps<IVerificationState>): React.ReactElement => (
            <>
              <FormTextInput
                formProps={formProps}
                inputType="phone"
                name="phone"
                label="Phone"
                inputPrefixText={formProps.values.phoneCode}
                isMandatory
                phoneFieldDropdownText={t('auth:countryRegion')}
                placeholder={t('yourNumber')}
                helpText={t('otpVerification')}
              />
              <FormButton
                // @ts-ignore
                onPress={formProps.handleSubmit}
                formProps={formProps}
                type="primary"
                title={buttonTitle}
                containerStyle={styles.formButtonStyle}
              />
            </>
          )}
        </Formik>
      </Screen>
    );
  }

  private onSubmit = async (
    values: IVerificationState,
    formActions: FormikHelpers<IVerificationState>
  ): Promise<void> => {
    formActions.setSubmitting(true);
    const {
      t,
      route: {
        params: { isFromLogin, userData, onCallback },
      },
      navigation,
    } = this.props;
    const { phone, phoneCode } = values;

    try {
      const formattedPhone = `${phoneCode}~${phone}`;
      const isPhoneUsed = await UserRepository.phoneExists(formattedPhone);
      if (isPhoneUsed.is_exists) {
        AlertHelper.error({ message: t('auth:phoneAlreadyExists') });
        return;
      }
      navigation.navigate(ScreensKeys.OTP, {
        type: OtpNavTypes.SocialMedia,
        title: isFromLogin ? t('loginOtp') : t('verifyNumber'),
        otpSentTo: phone,
        countryCode: phoneCode,
        socialUserData: userData,
        ...(onCallback && { onCallback }),
      });
    } catch (err) {
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(err.details) });
    }
  };

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private getDisplayStrings = (): { title: string; subTitle: string; message: string; buttonTitle: string } => {
    const {
      t,
      route: {
        params: { userData, isFromLogin },
      },
    } = this.props;
    const {
      provider,
      user: { first_name, email },
    } = userData;

    let title = t('signUpWithFacebook');
    const messageKey = isFromLogin ? 'enterNumberForProfileForLogin' : 'enterNumberForProfileForSignUp';

    if (isFromLogin) {
      if (provider === SocialAuthKeys.Google) {
        title = t('loginWithGoogle');
      } else if (provider === SocialAuthKeys.Apple) {
        title = t('loginWithApple');
      } else {
        title = t('loginWithFacebook');
      }
    }

    if (!isFromLogin) {
      switch (provider) {
        case SocialAuthKeys.Google:
          title = t('signUpWithGoogle');
          break;
        case SocialAuthKeys.Apple:
          title = t('signUpWithApple');
          break;
        default:
          title = t('signUpWithFacebook');
      }
    }

    return {
      title,
      subTitle: email,
      buttonTitle: isFromLogin ? t('common:login') : t('common:signUp'),
      message: t(`${messageKey}`, { givenName: first_name }),
    };
  };

  private formSchema = (): yup.ObjectSchema<{ phone: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      phone: yup.string().required(t('numberRequired')),
    });
  };
}

export default withTranslation(LocaleConstants.namespacesKey.auth)(MobileVerificationScreen);

const styles = StyleSheet.create({
  message: {
    color: theme.colors.darkTint3,
    marginBottom: 12,
  },
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
  container: {
    paddingVertical: theme.layout.screenPadding,
  },
});
