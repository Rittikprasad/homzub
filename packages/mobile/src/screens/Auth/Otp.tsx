import React, { ReactElement } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { StringUtils } from '@homzhub/common/src/utils/StringUtils';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { UserRepository } from '@homzhub/common/src/domain/repositories/UserRepository';
import { IUserTokens, StorageKeys, StorageService } from '@homzhub/common/src/services/storage/StorageService';
import { AnalyticsService } from '@homzhub/common/src/services/Analytics/AnalyticsService';
import { IRedirectionDetails } from '@homzhub/mobile/src/services/LinkingService';
import { NavigationService } from '@homzhub/mobile/src/services/NavigationService';
import { UserService } from '@homzhub/common/src/services/UserService';
import { CommonSelectors } from '@homzhub/common/src/modules/common/selectors';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { Text } from '@homzhub/common/src/components/atoms/Text';
import { OtpTimer } from '@homzhub/common/src/components/atoms/OtpTimer';
import { OtpInputs, OtpTypes } from '@homzhub/common/src/components/molecules/OtpInputs';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import { User } from '@homzhub/common/src/domain/models/User';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { NavigationScreenProps, OtpNavTypes, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import {
  IEmailLoginPayload,
  ILoginPayload,
  IOtpLoginPayload,
  LoginTypes,
} from '@homzhub/common/src/domain/repositories/interfaces';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { EventType } from '@homzhub/common/src/services/Analytics/EventType';

interface IStateProps {
  isLoading: boolean;
  redirectionDetails: IRedirectionDetails;
}

interface IDispatchProps {
  login: (payload: ILoginPayload) => void;
  loginSuccess: (data: IUserTokens) => void;
}

type libraryProps = NavigationScreenProps<AuthStackParamList, ScreensKeys.OTP> & WithTranslation;
type IProps = IDispatchProps & IStateProps & libraryProps;

interface IOtpState {
  error: boolean;
  emailOtp?: string;
  phoneOrEmailOtp?: string;
}

export class Otp extends React.PureComponent<IProps, IOtpState> {
  public state = {
    error: false,
    emailOtp: '',
    phoneOrEmailOtp: '',
  };

  public componentDidMount = async (): Promise<void> => {
    const {
      route: {
        params: { updateProfileCallback },
      },
    } = this.props;

    if (!updateProfileCallback) {
      await this.fetchOtp();
    }
  };

  public render = (): React.ReactNode => {
    const {
      t,
      isLoading,
      navigation: { goBack },
      route: { params },
    } = this.props;
    const title = params?.title ?? t('verifyNumber');
    const otpSentTo = params?.otpSentTo ?? '';
    const countryCode = params?.countryCode ?? '';

    return (
      <Screen
        headerProps={{ type: 'secondary', onIconPress: goBack }}
        pageHeaderProps={{
          contentTitle: title,
          contentSubTitle: t('enterOTP'),
          disableDivider: true,
        }}
        isLoading={isLoading}
        backgroundColor={theme.colors.white}
      >
        {this.renderOtpInputSection(`${countryCode} ${otpSentTo}`, OtpTypes.PhoneOrEmail)}
        {params.type === OtpNavTypes.UpdateProfileByEmailPhoneOtp
          ? this.renderOtpInputSection(params?.email || '', OtpTypes.Email)
          : null}
      </Screen>
    );
  };

  private renderOtpInputSection = (otpSentTo: string, otpType?: OtpTypes): ReactElement => {
    const {
      t,
      route: {
        params: { type },
      },
    } = this.props;
    const { error } = this.state;

    const toggleError = (): void => {
      this.toggleErrorState(false);
    };

    return (
      <>
        <View style={styles.numberContainer}>
          <Text type="small" textType="semiBold">
            {otpSentTo}
          </Text>
          <Icon
            name={icons.noteBook}
            size={16}
            color={theme.colors.active}
            style={styles.icon}
            onPress={this.onIconPress}
            testID="icnEdit"
          />
        </View>
        <OtpInputs
          error={type !== OtpNavTypes.UpdateProfileByEmailPhoneOtp && error ? t('otpError') : undefined}
          bubbleOtp={this.handleOtpVerification}
          toggleError={toggleError}
          otpType={otpType}
        />
        <OtpTimer onResentPress={this.fetchOtp} />
      </>
    );
  };

  private onIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private fetchOtp = async (): Promise<void> => {
    const {
      route: {
        params: { otpSentTo, countryCode },
      },
    } = this.props;

    try {
      if (StringUtils.isValidEmail(otpSentTo)) {
        await UserService.fetchEmailOtp(otpSentTo);
        return;
      }

      await UserService.fetchOtp(otpSentTo, countryCode);
    } catch (e) {
      AlertHelper.error({
        message: e.message,
      });
    }
  };

  private verifyOtp = async (otp: string): Promise<void> => {
    const {
      route: {
        params: { otpSentTo, countryCode, type },
      },
    } = this.props;

    if (type === OtpNavTypes.Login) {
      this.loginOtp(otp ?? '');
      return;
    }

    if (type === OtpNavTypes.SocialMedia) {
      await this.socialSignUp(otp);
      return;
    }

    try {
      await UserService.verifyOtp(otp, otpSentTo, countryCode);
      if (type === OtpNavTypes.SignUp) {
        await this.signUp();
      }
    } catch (e) {
      this.toggleErrorState(true);
    }
  };

  private signUp = async (): Promise<void> => {
    const {
      login,
      route: {
        params: { userData, onCallback },
      },
    } = this.props;

    if (!userData) {
      return;
    }

    try {
      await UserRepository.signUp(userData);
      const loginData: IEmailLoginPayload = {
        action: LoginTypes.EMAIL,
        payload: {
          email: userData.email,
          password: userData.password,
        },
      };
      const loginPayload: ILoginPayload = {
        data: loginData,
        is_referral: !!userData.signup_referral_code,
        is_from_signup: true,
        ...(onCallback && { callback: onCallback }),
        handleDynamicLink: this.handleDynamicLink,
      };
      login(loginPayload);
    } catch (e) {
      AlertHelper.error({ message: e.message });
    }
  };

  private handleDynamicLink = (): void => {
    const { redirectionDetails } = this.props;
    if (redirectionDetails.shouldRedirect && redirectionDetails.redirectionLink) {
      NavigationService.handleDynamicLinkNavigation(redirectionDetails).then();
    }
  };

  private socialSignUp = async (otp: string): Promise<void> => {
    const {
      loginSuccess,
      route: {
        params: { socialUserData, otpSentTo, countryCode },
      },
    } = this.props;

    if (!socialUserData || !otpSentTo || !countryCode) {
      return;
    }

    const trackData = {
      source: socialUserData.provider,
      email: socialUserData.user.email,
      phone_number: otpSentTo,
    };

    try {
      const data: User = await UserRepository.socialSignUp({
        provider: socialUserData.provider,
        id_token: socialUserData.idToken,
        otp,
        phone_code: countryCode,
        phone_number: otpSentTo,
      });

      const tokens = { refresh_token: data.refreshToken, access_token: data.accessToken };
      loginSuccess(tokens);
      await StorageService.set<IUserTokens>(StorageKeys.USER, tokens);

      AnalyticsService.track(EventType.SignupSuccess, trackData);
      // @ts-ignore
      AnalyticsService.setUser(ObjectMapper.deserialize(User, socialUserData.user));
    } catch (e) {
      AnalyticsService.track(EventType.SignupFailure, { ...trackData, error: e.message });
      AlertHelper.error({ message: e.message });
    }
  };

  private loginOtp = (otp: string): void => {
    const {
      login,
      route: {
        params: { otpSentTo, countryCode, onCallback },
      },
    } = this.props;

    const loginData: IOtpLoginPayload = {
      action: LoginTypes.OTP,
      payload: {
        phone_code: countryCode,
        phone_number: otpSentTo,
        otp,
      },
    };

    const loginPayload: ILoginPayload = {
      data: loginData,
      callback: onCallback,
      handleDynamicLink: this.handleDynamicLink,
    };
    login(loginPayload);
  };

  private toggleErrorState = (error: boolean): void => {
    this.setState({ error });
  };

  private handleOtpVerification = async (otp: string, otpType?: OtpTypes): Promise<void> => {
    const {
      route: {
        params: { updateProfileCallback, type },
      },
    } = this.props;

    if (!updateProfileCallback) {
      await this.verifyOtp(otp);
      return;
    }

    if (type === OtpNavTypes.UpdateProfileByEmailPhoneOtp) {
      this.setState(
        () => (otpType === OtpTypes.PhoneOrEmail ? { phoneOrEmailOtp: otp } : { emailOtp: otp }),
        () => {
          const { emailOtp, phoneOrEmailOtp } = this.state;
          if (emailOtp && phoneOrEmailOtp) {
            updateProfileCallback(phoneOrEmailOtp, emailOtp);
          }
        }
      );
      return;
    }

    updateProfileCallback(otp);
  };
}

export const mapStateToProps = (state: IState): IStateProps => {
  return {
    isLoading: UserSelector.getLoadingState(state),
    redirectionDetails: CommonSelectors.getRedirectionDetails(state) as IRedirectionDetails,
  };
};

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { login, loginSuccess } = UserActions;
  return bindActionCreators(
    {
      login,
      loginSuccess,
    },
    dispatch
  );
};

export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.auth)(Otp));

const styles = StyleSheet.create({
  numberContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  icon: {
    marginStart: 8,
  },
});
