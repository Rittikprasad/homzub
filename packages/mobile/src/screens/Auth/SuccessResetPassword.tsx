import React, { Component } from 'react';
import { CommonActions } from '@react-navigation/native';
import { StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { CommonParamList } from '@homzhub/mobile/src/navigation/Common';

export type Props = WithTranslation & NavigationScreenProps<CommonParamList, ScreensKeys.SuccessResetPassword>;

export class SuccessResetPassword extends Component<Props, {}> {
  public render(): React.ReactNode {
    const { t } = this.props;
    return (
      <Screen
        headerProps={{
          type: 'secondary',
          icon: icons.close,
          onIconPress: this.navigateToLogin,
        }}
        pageHeaderProps={{
          contentTitle: t('auth:successResetPassword'),
          contentSubTitle: t('auth:successResetPasswordDescription'),
          disableDivider: true,
        }}
        backgroundColor={theme.colors.white}
      >
        <Button
          type="primary"
          title={t('common:login')}
          onPress={this.navigateToLogin}
          containerStyle={styles.button}
          testID="btnLogin"
        />
      </Screen>
    );
  }

  public navigateToLogin = (): void => {
    const { navigation } = this.props;

    navigation.dispatch(
      CommonActions.reset({
        index: 0,
        routes: [{ name: ScreensKeys.Login }],
      })
    );
  };
}

export default withTranslation()(SuccessResetPassword);

const styles = StyleSheet.create({
  button: {
    flex: 0,
  },
});
