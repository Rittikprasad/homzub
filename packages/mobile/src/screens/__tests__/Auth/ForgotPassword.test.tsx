import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ForgotPassword } from '@homzhub/mobile/src/screens/Auth/ForgotPassword';

const mock = jest.fn();

describe('Forgot Password Screen', () => {
  let component: any;
  let props: any;

  beforeEach(() => {
    props = {
      navigation: {
        navigate: mock,
        goBack: mock,
      },
    };
    component = shallow(
      <ForgotPassword
        {...props}
        t={(key: string): string => key}
        route={{
          params: {
            onCallback: mock,
          },
        }}
      />
    );
  });

  it('should render forgot password screen', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should navigate to login screen', () => {
    component.find('[testID="txtLogin"]').prop('onPress')();
    expect(mock).toHaveBeenCalled();
  });
});
