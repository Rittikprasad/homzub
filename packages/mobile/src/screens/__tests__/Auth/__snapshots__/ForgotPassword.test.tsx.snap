// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forgot Password Screen should render forgot password screen 1`] = `
<Screen
  backgroundColor="#FFFFFF"
  headerProps={
    Object {
      "icon": "close",
      "onIconPress": [Function],
      "type": "secondary",
    }
  }
  pageHeaderProps={
    Object {
      "contentSubTitle": "auth:forgotPasswordDescription",
      "contentTitle": "auth:forgotPassword",
      "disableDivider": true,
    }
  }
>
  <Formik
    initialValues={
      Object {
        "email": "",
        "isRequestDisable": false,
      }
    }
    onSubmit={[Function]}
    validate={[Function]}
  >
    <Component />
  </Formik>
  <Text
    onPress={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "color": "#0084F8",
      }
    }
    testID="txtLogin"
    textType="semiBold"
    type="small"
  >
    auth:backToLogin
  </Text>
</Screen>
`;
